File: appli-web/src/app/features/nutrition/micros.ts
/**
 * Vitamines & minéraux (pack essentiel ~10, D11 étendu 2026-06-13) — helpers d'affichage info.
 *
 * v1 : affichage seulement (pas de cibles). Minéraux + vitamine C en mg, vitamines D/B12/A en µg.
 * Les valeurs sont per-100 g (snapshot/catalogue). Pures, testables sans Angular.
 */

/** Sous-ensemble structurel d'un Food/MealEntry/OffProduct portant les 10 micros. */
export interface MicroNutrients {
  ironPer100g: number | null;
  calciumPer100g: number | null;
  magnesiumPer100g: number | null;
  zincPer100g: number | null;
  potassiumPer100g: number | null;
  sodiumPer100g: number | null;
  vitaminCPer100g: number | null;
  vitaminDPer100g: number | null;
  vitaminB12Per100g: number | null;
  vitaminAPer100g: number | null;
}

interface MicroSpec {
  key: keyof MicroNutrients;
  label: string;
  /** Abréviation compacte pour les lignes denses (symbole chimique pour les minéraux). */
  short: string;
  unit: 'mg' | 'µg';
}

/** Ordre d'affichage : minéraux puis vitamines (cohérent avec le cadrage de la tâche). */
const MICRO_SPECS: readonly MicroSpec[] = [
  { key: 'ironPer100g', label: 'Fer', short: 'Fe', unit: 'mg' },
  { key: 'calciumPer100g', label: 'Calcium', short: 'Ca', unit: 'mg' },
  { key: 'magnesiumPer100g', label: 'Magnésium', short: 'Mg', unit: 'mg' },
  { key: 'zincPer100g', label: 'Zinc', short: 'Zn', unit: 'mg' },
  { key: 'potassiumPer100g', label: 'Potassium', short: 'K', unit: 'mg' },
  { key: 'sodiumPer100g', label: 'Sodium', short: 'Na', unit: 'mg' },
  { key: 'vitaminCPer100g', label: 'Vit. C', short: 'Vit C', unit: 'mg' },
  { key: 'vitaminDPer100g', label: 'Vit. D', short: 'Vit D', unit: 'µg' },
  { key: 'vitaminB12Per100g', label: 'Vit. B12', short: 'Vit B12', unit: 'µg' },
  { key: 'vitaminAPer100g', label: 'Vit. A', short: 'Vit A', unit: 'µg' },
];

export interface MicroDisplayItem {
  label: string;
  /** Abréviation compacte (Fe, Ca, Mg… ; « Vit C » pour les vitamines). */
  short: string;
  value: number;
  unit: string;
}

/** Clés des 10 micros, ordre minéraux puis vitamines (= ordre d'affichage des barres/anneaux). */
export const MICRO_KEYS: readonly (keyof MicroNutrients)[] = MICRO_SPECS.map((s) => s.key);

/** Abréviation compacte par clé : symbole chimique pour les minéraux (Fe, Ca, Mg, Zn, K, Na), « Vit C/D/B12/A »
 *  pour les vitamines. Sert aux libellés denses (parts de donut, lignes compactes). */
export const MICRO_SHORT: Record<keyof MicroNutrients, string> = MICRO_SPECS.reduce(
  (acc, s) => ({ ...acc, [s.key]: s.short }),
  {} as Record<keyof MicroNutrients, string>,
);

/** Cumul jour d'un micro pour chacune des 10 clés (valeurs absolues, jamais null — 0 par défaut). */
export type MicroTotals = Record<keyof MicroNutrients, number>;

export const ZERO_MICRO_TOTALS: MicroTotals = MICRO_SPECS.reduce(
  (acc, s) => ({ ...acc, [s.key]: 0 }),
  {} as MicroTotals,
);

/** Famille d'un micro (teinte d'affichage par famille, pas de couleur cible imposée). */
export type MicroFamily = 'mineral' | 'vitamin';

/**
 * Cible d'un micro. Pour 9 des 10 micros = VNR officielle UE (règlement 1169/2011, annexe XIII),
 * objectif à atteindre. Pour le Sodium = repère LIMITE (plafond 2000 mg, repère OMS), pas un
 * objectif : à afficher comme plafond, en alerte si dépassé (`isLimit`).
 */
export interface MicroTarget {
  key: keyof MicroNutrients;
  label: string;
  unit: 'mg' | 'µg';
  family: MicroFamily;
  /** VNR (objectif) ou plafond (Sodium). */
  target: number;
  /** true => repère plafond (Sodium) : barre d'alerte si dépassé, pas un objectif à remplir. */
  isLimit: boolean;
}

/** Cibles VNR UE + repère plafond Sodium (constantes fixes, ordre minéraux puis vitamines). */
export const MICRO_TARGETS: readonly MicroTarget[] = [
  { key: 'ironPer100g', label: 'Fer', unit: 'mg', family: 'mineral', target: 14, isLimit: false },
  { key: 'calciumPer100g', label: 'Calcium', unit: 'mg', family: 'mineral', target: 800, isLimit: false },
  { key: 'magnesiumPer100g', label: 'Magnésium', unit: 'mg', family: 'mineral', target: 375, isLimit: false },
  { key: 'zincPer100g', label: 'Zinc', unit: 'mg', family: 'mineral', target: 10, isLimit: false },
  { key: 'potassiumPer100g', label: 'Potassium', unit: 'mg', family: 'mineral', target: 2000, isLimit: false },
  { key: 'sodiumPer100g', label: 'Sodium', unit: 'mg', family: 'mineral', target: 2000, isLimit: true },
  { key: 'vitaminCPer100g', label: 'Vit. C', unit: 'mg', family: 'vitamin', target: 80, isLimit: false },
  { key: 'vitaminDPer100g', label: 'Vit. D', unit: 'µg', family: 'vitamin', target: 5, isLimit: false },
  { key: 'vitaminB12Per100g', label: 'Vit. B12', unit: 'µg', family: 'vitamin', target: 2.5, isLimit: false },
  { key: 'vitaminAPer100g', label: 'Vit. A', unit: 'µg', family: 'vitamin', target: 800, isLimit: false },
];

/** Ligne d'affichage d'un micro vs sa cible (barre ou anneau), dérivée des cumuls du jour. */
export interface MicroRow extends MicroTarget {
  value: number;
  /** Avancement borné 0..1 (pour la barre / l'anneau). */
  progress: number;
  /** Sodium seulement : cumul strictement au-dessus du plafond (déclenche la couleur d'alerte). */
  exceeded: boolean;
}

/** Construit les 10 lignes micros (cumul vs VNR / plafond) — pure, testable sans Angular. */
export function microRows(totals: MicroTotals): MicroRow[] {
  return MICRO_TARGETS.map((s) => {
    const value = totals[s.key];
    const ratio = s.target > 0 ? value / s.target : 0;
    return {
      ...s,
      value,
      progress: Math.max(0, Math.min(1, ratio)),
      exceeded: s.isLimit && value > s.target,
    };
  });
}

/**
 * Micros présents (valeur non nulle), arrondis à 1 décimale. `factor` met les valeurs à l'échelle
 * d'une quantité consommée (ex. `quantityG / 100` pour une entry du journal) ; 1 par défaut =
 * per-100 g. Donnée structurée canonique : l'affichage COLORÉ par famille passe par `microLineItems`
 * (micro-colors.ts) — aucune couleur n'est décidée ici (séparation donnée / code couleur).
 */
export function presentMicros(m: MicroNutrients, factor = 1): MicroDisplayItem[] {
  return MICRO_SPECS.filter((s) => m[s.key] != null).map((s) => ({
    label: s.label,
    short: s.short,
    value: Math.round((m[s.key] as number) * factor * 10) / 10,
    unit: s.unit,
  }));
}

File: appli-android/app/src/main/java/com/example/sportapp/feature/nutrition/domain/Micros.kt
package com.example.sportapp.feature.nutrition.domain

import com.example.sportapp.core.data.model.Food
import com.example.sportapp.core.data.model.MealEntry
import kotlin.math.roundToInt

/**
 * Vitamines & minéraux (pack essentiel 10, D11 étendu) — logique pure, testable
 * sans Compose ni Android. Miroir Android de `appli-web/.../nutrition/micros.ts`.
 *
 * v1 : affichage seulement (pas de cibles éditables). Cibles = VNR UE
 * (règlement 1169/2011) pour 9 micros ; Sodium = repère PLAFOND (2000 mg, repère
 * OMS). Valeurs snapshot per-100 g sur la MealEntry, mises à l'échelle de la
 * quantité consommée (× quantityG / 100). Les labels d'affichage vivent dans
 * `strings.xml` (politique 18) et sont résolus côté composable via la clé enum.
 */

/** Famille d'un micronutriment (teinte d'affichage par famille, pas par %VNR). */
enum class MicroFamily { MINERAL, VITAMIN }

/** Unité d'affichage d'un micronutriment. */
enum class MicroUnit(val symbol: String) { MG("mg"), UG("µg") }

/**
 * Les 10 micros suivis. UPPER_CASE (politique 11). Chaque clé porte sa cible
 * (VNR ou plafond), son unité, sa famille, et le flag `isLimit` (Sodium = repère
 * plafond, pas objectif). Ordre = minéraux puis vitamines (= ordre d'affichage).
 */
enum class MicroKey(
    val family: MicroFamily,
    val unit: MicroUnit,
    /** VNR (objectif à atteindre) ou plafond (Sodium). */
    val target: Float,
    /** true => repère plafond (Sodium) : alerte si dépassé, pas un objectif. */
    val isLimit: Boolean,
) {
    IRON(MicroFamily.MINERAL, MicroUnit.MG, 14f, false),
    CALCIUM(MicroFamily.MINERAL, MicroUnit.MG, 800f, false),
    MAGNESIUM(MicroFamily.MINERAL, MicroUnit.MG, 375f, false),
    ZINC(MicroFamily.MINERAL, MicroUnit.MG, 10f, false),
    POTASSIUM(MicroFamily.MINERAL, MicroUnit.MG, 2000f, false),
    SODIUM(MicroFamily.MINERAL, MicroUnit.MG, 2000f, true),
    VITAMIN_C(MicroFamily.VITAMIN, MicroUnit.MG, 80f, false),
    VITAMIN_D(MicroFamily.VITAMIN, MicroUnit.UG, 5f, false),
    VITAMIN_B12(MicroFamily.VITAMIN, MicroUnit.UG, 2.5f, false),
    VITAMIN_A(MicroFamily.VITAMIN, MicroUnit.UG, 800f, false),
}

/** Valeur per-100 g d'un micro sur une entry (null si absente de la source). */
fun MealEntry.microPer100g(key: MicroKey): Float? = when (key) {
    MicroKey.IRON -> ironPer100g
    MicroKey.CALCIUM -> calciumPer100g
    MicroKey.MAGNESIUM -> magnesiumPer100g
    MicroKey.ZINC -> zincPer100g
    MicroKey.POTASSIUM -> potassiumPer100g
    MicroKey.SODIUM -> sodiumPer100g
    MicroKey.VITAMIN_C -> vitaminCPer100g
    MicroKey.VITAMIN_D -> vitaminDPer100g
    MicroKey.VITAMIN_B12 -> vitaminB12Per100g
    MicroKey.VITAMIN_A -> vitaminAPer100g
}

/** Valeur per-100 g d'un micro sur un aliment du catalogue (null si absente de la source). */
fun Food.microPer100g(key: MicroKey): Float? = when (key) {
    MicroKey.IRON -> ironPer100g
    MicroKey.CALCIUM -> calciumPer100g
    MicroKey.MAGNESIUM -> magnesiumPer100g
    MicroKey.ZINC -> zincPer100g
    MicroKey.POTASSIUM -> potassiumPer100g
    MicroKey.SODIUM -> sodiumPer100g
    MicroKey.VITAMIN_C -> vitaminCPer100g
    MicroKey.VITAMIN_D -> vitaminDPer100g
    MicroKey.VITAMIN_B12 -> vitaminB12Per100g
    MicroKey.VITAMIN_A -> vitaminAPer100g
}

/**
 * Cumul jour des 10 micros depuis les snapshots per-100 g des entries
 * (total = per100g × quantityG / 100, null traité comme 0). Pur.
 */
fun sumMicroTotals(entries: List<MealEntry>): Map<MicroKey, Float> {
    val acc = MicroKey.entries.associateWith { 0f }.toMutableMap()
    for (e in entries) {
        val factor = e.quantityG / 100f
        for (k in MicroKey.entries) {
            acc[k] = (acc[k] ?: 0f) + (e.microPer100g(k) ?: 0f) * factor
        }
    }
    return acc
}

/** Ligne d'affichage d'un micro vs sa cible (barre), dérivée des cumuls du jour. */
data class MicroRow(
    val key: MicroKey,
    val value: Float,
    /** Avancement borné 0..1 (pour la barre). */
    val progress: Float,
    /** Sodium seulement : cumul strictement au-dessus du plafond (couleur d'alerte). */
    val exceeded: Boolean,
)

/** Construit les 10 lignes micros (cumul vs VNR / plafond) — pur, testable. */
fun microRows(totals: Map<MicroKey, Float>): List<MicroRow> =
    MicroKey.entries.map { key ->
        val value = totals[key] ?: 0f
        val ratio = if (key.target > 0f) value / key.target else 0f
        MicroRow(
            key = key,
            value = value,
            progress = ratio.coerceIn(0f, 1f),
            exceeded = key.isLimit && value > key.target,
        )
    }

/**
 * Abréviation compacte d'un micro pour les lignes denses (symbole chimique pour les minéraux,
 * « Vit X » pour les vitamines). Symboles universels → non traduits (comme `short` côté web).
 */
val MicroKey.short: String
    get() = when (this) {
        MicroKey.IRON -> "Fe"
        MicroKey.CALCIUM -> "Ca"
        MicroKey.MAGNESIUM -> "Mg"
        MicroKey.ZINC -> "Zn"
        MicroKey.POTASSIUM -> "K"
        MicroKey.SODIUM -> "Na"
        MicroKey.VITAMIN_C -> "Vit C"
        MicroKey.VITAMIN_D -> "Vit D"
        MicroKey.VITAMIN_B12 -> "Vit B12"
        MicroKey.VITAMIN_A -> "Vit A"
    }

/** Un micro présent (valeur non nulle) prêt à afficher en ligne compacte, coloré par famille. */
data class MicroLineItem(
    val short: String,
    /** Valeur per-100 g arrondie à 1 décimale. */
    val value: Float,
    val unit: String,
    val family: MicroFamily,
)

/**
 * Micros CONSOMMÉS (mis à l'échelle de la quantité) présents d'une entry du journal,
 * en abréviations compactes — pour le dépli micros d'une `MacroEntryRow` (miroir du
 * `entryRow` web). Arrondi à 1 décimale. Pur.
 */
fun MealEntry.consumedMicroLineItems(): List<MicroLineItem> =
    MicroKey.entries.mapNotNull { key ->
        val per100 = microPer100g(key) ?: return@mapNotNull null
        val consumed = per100 * quantityG / 100f
        if (consumed <= 0f) return@mapNotNull null
        MicroLineItem(
            short = key.short,
            value = (consumed * 10f).roundToInt() / 10f,
            unit = key.unit.symbol,
            family = key.family,
        )
    }

/**
 * Micros présents (valeur non nulle) d'un aliment du catalogue, en abréviations compactes, pour la
 * ligne micros d'une row — miroir de `microLineItems` (web). Per-100 g arrondi à 1 décimale. Pur.
 */
fun Food.microLineItems(): List<MicroLineItem> =
    MicroKey.entries.mapNotNull { key ->
        val v = microPer100g(key) ?: return@mapNotNull null
        MicroLineItem(
            short = key.short,
            value = (v * 10f).roundToInt() / 10f,
            unit = key.unit.symbol,
            family = key.family,
        )
    }
